### ISSUE TITLE
Office missing a few days’ transactions

### ISSUE THREAD

**Issue by scannerdarkly** [CORE-POS/IS4C#966]
Due to a network configuration issue (since fixed), our backend is missing a few days’ transactions from mid-January. It has transactions from both before and after though.
 
Our lanes still have the missing transactions sitting in their databases (i.e. no LaneTrimTask has been run on them). Is there a safe, efficient way to transfer these transactions to the backend?

* Version of CORE?
1.10-dev (lanes)
2.7 (office)

* Issue with Office, Lane, or both?
both

* Is this [mostly] a bug report, feature request, or question?
question

**Comment by flathat** [comment]
You could just copy (`INSERT`) them to `dtransactions` on the lane and then the next time you log in or finish a transaction they will be uploaded (a significant delay in the login or transaction end will be apparent).  Check that the column sets in the two tables are the same; if they aren't you'll need to enumerate the columns in `INSERT INTO dtransactions SELECT ...`.

Then you can wait for the regular `TransArchiveTask` process to move the data to `transarchive` etc, or schedule a special run to handle the transactions prior to today.

**Comment by gohanman** [comment]
That sounds strange. Transaction data accumulates on the lane in `translog.dtransactions` and isn't cleared until the table's content is successfully copied to the server. There's no daily/periodic/automatic mechanism so I'm not sure how you'd get later data syncing again.

Coping data back into the lane's `translog.dtransactions` table should ship it to the server again, but without knowing what went wrong in the first place I can't say for certain this wouldn't create duplicate data on the server.

**Comment by scannerdarkly** [comment]
Looking at one of the lanes right now: `localtrans` has data going back to December 14th. `localtranstoday` has today’s transactions so far. `dtransactions` and `localtransarchive` are both empty. Is all this normal?

**Comment by gohanman** [comment]
Yes. `localtransarchive` AFAIK has never been used for anything. `dtransactions` is emptied each time its contents are successfully copied to the server. Typically that happens following every transaction but in the event that connectivity fails the lane's `dtransactions` table will keep accumulating data until connectivity is restored. What's perplexing here is:

* It shouldn't be possible for transaction data to reach the server without passing through `dtransactions`  on the lane first.
* It shouldn't be possible for data to be deleted from the lane's `dtransactions` table unless that data has been copied to the server.

This setup has been the same for as long as I know and certainly predates CORE. I can't remember any situation where these invariants didn't hold.

**Comment by scannerdarkly** [comment]
What process copies the data over? Perhaps I can emulate that process, checking before each copy in order to avoid duplications...

**Comment by scannerdarkly** [comment]
Is `dtransactions` normally filled from `localtrans`? Perhaps the data was never put into `dtransactions` in the first place — though I don’t know how that would happen either!

**Comment by gohanman** [comment]
> Is dtransactions normally filled from localtrans? Perhaps the data was never put into dtransactions in the first place — though I don’t know how that would happen either!

No. `localtrans`, `localtranstoday`, and `dtransactions` are all populated in [Database::rotateTempData](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/lib/Database.php#L800-L839). Copying data into `dtransactions` is what determines whether the function succeeded. If the copy fails the current transaction [won't end](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/lib/TransRecord.php#L788-L790)

> What process copies the data over? Perhaps I can emulate that process, checking before each copy in order to avoid duplications...

[Database::uploadtoServer](https://github.com/CORE-POS/IS4C/blob/master/pos/is4c-nf/lib/Database.php#L438-L491) handles it. The end of transaction and on logging in are the most prominent places this gets called.

**Comment by scannerdarkly** [comment]
Okay, thanks. I’ll dive into this this week!

**Comment by scannerdarkly** [comment]
Okay, I manually copied `localtrans` to `dtransactions` (on the lane), then logged in, which immediately ran `Database::uploadtoServer()`, which in turn ran `SQLManager::transfer()`. I had to rewrite `SQLManager::transfer()` to handle the size of this data — I’ll submit that code in a bit.

The lane’s `dtransactions` is now empty, but `localtrans` is still full. Is that dangerous in any way? (I do see that `LaneTrimTask::run()` deletes data in `localtrans` that’s over 30 days old.)
